cmd/launcher/secure_enclave_darwin.go

```diff
@@ -0,0 +1,163 @@
+		}
+
+		serverPubKeys[string(pubB64Der)] = key
+	}
+
+	switch args[0] {
+	case secureenclavesigner.CreateKeyCmd:
+		return createSecureEnclaveKey(args[1])
+
+	case secureenclavesigner.SignCmd:
+		return signWithSecureEnclave(args[1])
+
+	default:
+		return fmt.Errorf("unknown command %s", args[0])
+	}
+}
+
+func createSecureEnclaveKey(requestB64 string) error {
+	b, err := base64.StdEncoding.DecodeString(requestB64)
+	if err != nil {
+		return fmt.Errorf("decoding b64 request: %w", err)
+	}
+
+	var request secureenclavesigner.Request
+	if err := msgpack.Unmarshal(b, &request); err != nil {
+		return fmt.Errorf("unmarshaling msgpack request: %w", err)
+	}
+
+	if err := verifySecureEnclaveChallenge(request); err != nil {
+		return fmt.Errorf("verifying challenge: %w", err)
+	}
+
+	secureEnclavePubKey, err := secureenclave.CreateKey()
+	if err != nil {
+		return fmt.Errorf("creating secure enclave key: %w", err)
+	}
+
+	secureEnclavePubDer, err := echelper.PublicEcdsaToB64Der(secureEnclavePubKey)
+	if err != nil {
+		return fmt.Errorf("marshalling public key to der: %w", err)
+	}
+
+	// write results to stdout since so that parent process can use
+	fmt.Println(string(secureEnclavePubDer))
+	return nil
+}
+
+func signWithSecureEnclave(signRequestB64 string) error {
+	b, err := base64.StdEncoding.DecodeString(signRequestB64)
+	if err != nil {
+		return fmt.Errorf("decoding b64 sign request: %w", err)
+	}
+
+	var signRequest secureenclavesigner.SignRequest
+	if err := msgpack.Unmarshal(b, &signRequest); err != nil {
+		return fmt.Errorf("unmarshaling msgpack sign request: %w", err)
+	}
+
+	if err := verifySecureEnclaveChallenge(signRequest.Request); err != nil {
+		return fmt.Errorf("verifying challenge: %w", err)
+	}
+
+	secureEnclavePubKey, err := echelper.PublicB64DerToEcdsaKey(signRequest.SecureEnclavePubKey)
+	if err != nil {
+		return fmt.Errorf("marshalling b64 der to public key: %w", err)
+	}
+
+	seSigner, err := secureenclave.New(secureEnclavePubKey)
+	if err != nil {
+		return fmt.Errorf("creating secure enclave signer: %w", err)
+	}
+
+	sig, err := seSigner.Sign(rand.Reader, signRequest.Digest, crypto.SHA256)
+	if err != nil {
+		return fmt.Errorf("signing request: %w", err)
+	}
+
+	// write results to stdout since so that parent process can use
+	fmt.Print(base64.StdEncoding.EncodeToString(sig))
+	return nil
+}
+
+func verifySecureEnclaveChallenge(request secureenclavesigner.Request) error {
+	c, err := challenge.UnmarshalChallenge(request.Challenge)
+	if err != nil {
+		return fmt.Errorf("unmarshaling challenge: %w", err)
+	}
+
+	serverPubKey, ok := serverPubKeys[string(request.ServerPubKey)]
+	if !ok {
+		return errors.New("server public key not found")
+	}
+
+	if err := c.Verify(*serverPubKey); err != nil {
+		return fmt.Errorf("verifying challenge: %w", err)
+	}
+
+	// Check the timestamp, this prevents people from saving a challenge and then
+	// reusing it a bunch. However, it will fail if the clocks are too far out of sync.
```

Comment:
If I'm understanding correctly we need the server to adjust the time stamp to the the machine time? I think this could be handled in root launcher before we get here if I'm understanding correctly.

Reply:
Handling it in the root launcher signing makes sense.